Http 204 and alias of optional

iamdanfox opened the pull request:
## Before this PR

@pkoenig10 reported the problem (fixes https://github.com/palantir/dialogue/issues/573). This is high-priority because it's a regression compared to c-j-r's deserialization.

## After this PR
==COMMIT_MSG==
When used with conjure-java generated java bindings, dialogue can now deserialize HTTP 204 into `FooAlias.of(Optional.empty())`. Previously it would throw.
==COMMIT_MSG==

## Possible downsides?
<!-- Please describe any way users could be negatively affected by this PR. -->

iamdanfox committed:
Repro alias of optional problems

carterkozak commented on dialogue-client-verifier/src/test/java/com/palantir/verification/IntegrationTest.java:
```diff
@@ -120,6 +138,13 @@ public void stream_3_gigabytes() throws IOException {
 
+    private void set204Response() {
+        undertowHandler = exchange -> {
+            exchange.getResponseHeaders().put(Headers.CONTENT_TYPE, "application/json");
```
This isn't valid, if there's no content it shouldn't have a type!

iamdanfox committed:
JacksonEmptyContainerLoader

iamdanfox committed:
ConjureBodySerDe stores a suppliers.memoize thingy

iamdanfox committed:
Add generated changelog entries

iamdanfox committed:
Merge remote-tracking branch 'origin/develop' into dfox/alias-of-optional

iamdanfox committed:
Smaller diff

carterkozak committed:
Update JacksonEmptyContainerLoader to use Type, not Class

This way we can avoid losing context.

ferozco commented on dialogue-serde/src/test/java/com/palantir/conjure/java/dialogue/serde/JacksonEmptyContainerLoaderTest.java:
```diff
@@ -0,0 +1,228 @@
+                .isEqualTo(AliasAlias1.of(Alias1.of(Optional.empty())));
+    }
+
+    @Generated("com.palantir.conjure.java.types.AliasGenerator")
```
can we avoid checking in generated code? We already generate bindings in other places

iamdanfox commented on dialogue-serde/src/main/java/com/palantir/conjure/java/dialogue/serde/ConjureBodySerDe.java:
```diff
@@ -218,11 +218,10 @@ public T deserialize(Response response) {
-                    return TypeMarkers.getEmptyOptional(token);
+                    // TODO(dfox): what if we get a 204 for a non-optional type???
+                    // TODO(dfox): support http200 & body=null
+                    // TODO(dfox): what if we were expecting an empty list but got {}?
```
adding extra strict tests upstream to the verifier: https://github.com/palantir/conjure-verification/pull/330

ferozco commented on dialogue-serde/src/main/java/com/palantir/conjure/java/dialogue/serde/ConjureBodySerDe.java:
```diff
@@ -50,6 +51,7 @@
     private final List<Encoding> encodingsSortedByWeight;
     private final ErrorDecoder errorDecoder;
     private final Encoding defaultEncoding;
+    private final EmptyContainerDeserializer empties;
```
s/empties/emptyBodyDecoder/

ferozco commented on dialogue-serde/src/main/java/com/palantir/conjure/java/dialogue/serde/ConjureBodySerDe.java:
```diff
@@ -195,17 +193,19 @@ public void close() {
             this.errorDecoder = errorDecoder;
-            this.token = token;
-            this.isOptionalType = TypeMarkers.isOptional(token);
+            this.emptyInstance = Suppliers.memoize(() -> empty.getEmptyInstance(token)); // TODO(dfox): save exception?
```
I think by memoizing we will lose the exception if mulitple calls receive a 204 for a non empty type. can we construct the instance at response time?

iamdanfox replied:
I'd prefer not to construct the instance every time because we always expect to get the same answer

ferozco replied:
I was wrong, memoizing won't interfere with the exception so this works well

ferozco commented on dialogue-serde/src/main/java/com/palantir/conjure/java/dialogue/serde/ConjureBodySerDe.java:
```diff
@@ -218,11 +218,10 @@ public T deserialize(Response response) {
-                                "Unable to deserialize non-optional response type from 204", SafeArg.of("type", token));
-                    }
-                    return TypeMarkers.getEmptyOptional(token);
+                    // TODO(dfox): what if we get a 204 for a non-optional type???
```
should be handled by moving the instance construction

ferozco commented on dialogue-serde/src/main/java/com/palantir/conjure/java/dialogue/serde/ConjureBodySerDe.java:
```diff
@@ -218,11 +218,10 @@ public T deserialize(Response response) {
-                    }
-                    return TypeMarkers.getEmptyOptional(token);
+                    // TODO(dfox): what if we get a 204 for a non-optional type???
+                    // TODO(dfox): support http200 & body=null
```
These last two cases are handled by the regular deserializer

carterkozak replied:
In most cases that's correct, unfortunately null->empty collection won't work that way because we can't provide field annotations.

ferozco commented on dialogue-serde/src/main/java/com/palantir/conjure/java/dialogue/serde/Encodings.java:
```diff
@@ -37,6 +39,9 @@
 
     private Encodings() {}
 
+    private static final Supplier<ObjectMapper> JSON_MAPPER =
```
why memoize? can't we just have a static field?

carterkozak replied:
imo it's a bit cleaner to avoid work until it's requested, we don't want expensive class loading. When failures occur, it's much nicer to have a stack trace through a call tree than a class init error.

iamdanfox replied:
In addition to the stacktraces, there's also the perf thing: if everyone did big work in their static fields, then witchcraft startup would get slower and slower.  If we all use suppliers.memoize then the world starts up fast

carterkozak replied:
We use a variation of this in an internal project to memoize encoding.serializer/deserializer creation in conjure-undertow, to avoid relatively expensive ObjectReader/ObjectWriter creation until the first time they're needed. Might be worth implementing by default in both dialogue and conjure-undertow.

ferozco replied:
Ok, its pretty ubiquitous to have static mappers. Perhaps we should write an errorprone rule to try to avoid it. IMO its not that big of deal though

ferozco commented on dialogue-serde/src/main/java/com/palantir/conjure/java/dialogue/serde/JacksonEmptyContainerLoader.java:
```diff
@@ -0,0 +1,153 @@
+    @Override
+    @SuppressWarnings("unchecked")
+    public <T> T getEmptyInstance(TypeMarker<T> token) {
+        Class<?> clazz = getClass(token.getType());
```
Could we do something like:
```
ObjectReader objectReader = mapper.readerFor(mapper.constructType(type.getType()));
return objectReader.readValue(mapper.nullNode());
```

carterkozak replied:
I created a PR into this branch which does basically that

carterkozak replied:
Link #592

carterkozak committed:
nullNode + tests

iamdanfox committed:
Don't check in generated code

iamdanfox committed:
Merge remote-tracking branch 'origin/ckozak/use_type_not_class' into dfox/alias-of-optional

iamdanfox committed:
s/empties/emptyBodyDecoder/

carterkozak committed:
Refactor optional deserialization to avoid memoization

While I think it may be helpful to memoize encoding bits, it's
odd that we memoize only some pieces.

carterkozak committed:
tests

carterkozak committed:
tryGetEmptyInstance

iamdanfox committed:
JUnit5 everywhere

iamdanfox committed:
Merge remote-tracking branch 'origin/ckozak/avoid_memoize' into dfox/alias-of-optional

carterkozak commented:
fortunately the builds aren't passing, so we've only burned a tag.

iamdanfox committed:
upgrade conjure-verification to get more test cases

iamdanfox commented on dialogue-serde/src/main/java/com/palantir/conjure/java/dialogue/serde/JacksonEmptyContainerLoader.java:
```diff
@@ -0,0 +1,158 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+final class JacksonEmptyContainerLoader implements EmptyContainerDeserializer {
```
Linking to the original impl: https://github.com/palantir/conjure-java-runtime/blob/45d6c9fa668f000bcd86e9f8003789db5d70d843/conjure-java-jaxrs-client/src/main/java/com/palantir/conjure/java/client/jaxrs/feignimpl/EmptyContainerDecoder.java#L79

iamdanfox commented:
:+1:

svc-autorelease commented:
Released [1.5.1](https://github.com/palantir/dialogue/releases/tag/1.5.1)
